pkg/gornir/gornir.go

```diff
@@ -16,6 +17,41 @@ type Gornir struct {
 	Logger    Logger     // Logger for the object
 }
 
+// New is a Gornir constructor. It is currently no different that new,
+// however is a placeholder for any future defaults.
+func New() *Gornir {
+	return new(Gornir)
+}
+
+// InventoryPlugin is an Inventory Source
+type InventoryPlugin interface {
+	Create() (Inventory, error)
+}
+
+// WithInventory creates a new Gornir with an Inventory.
+func (gr *Gornir) WithInventory(inv Inventory) *Gornir {
+	return &Gornir{
+		Logger:    gr.Logger,
+		Inventory: &inv,
+	}
+}
+
+// WithFilter creates a new Gornir with a filtered Inventory.
+func (gr *Gornir) WithFilter(f FilterFunc) *Gornir {
```

Comment:
I'd prefer calling this just `Filter`, mostly for consistency; `WithSomething` attaches that `Something` to the returned object while `Filter` applies a transformation (the inventory is still the same, it just changed due to the filter function).

Reply:
I want with `WithFilter` for a couple of reasons:

1. `Filter` takes a context right now, which I would prefer to avoid until deemed necessary.

2. `original.WithFilter(filter)` returns a completely new `Gornir` **with** a filter applied over the contents of `original`. `original` remains unmodified, so we didn't filter it.

3. Consistency across options; if starts with With, it's an option :-)

Reply:
I am good removing the context for now (I'd remove it in Inventory.Filter and FilterFunc too for consistency) but I am not sure `WithFilter` is still a consistent name here as it does something different. The other `With` methods receive attributes that are set in the new object (or options as you say). On the other hand, `WithFilter` doesn't set an attribute, it applies a transformation on an existing attribute.

Another way of looking at this is that `WithInventory` and `WithLogger` (I also want to add `WithRunner` so it doesn't have to be passed to `Gornir.Run` each time but no need to do it in this PR) can be called in any order, that's not true for `WithFilter`, again, because it doesn't set an attribute, it applies a transformation on an attribute.

And if that wasn't enough you can basically realize that `WithFilter` is just a shorthand for `gr.WithInventory(gr.Inventory.Filter(filterFunc))`. `Filter` is also consistent with `Inventory.Filter`.

Anyway, don't want to block this PR on this discussion so if you are still not convinced we can merge it anyway after removing the context everywhere and rehash it on a different PR :)

Reply:
I'll remove context from `Filter` (and `Inventory.Filter`, `FilterFunc`), plus get rid of `WithFilter`, then merge.

I'll add `WithRunner` in another PR.